### PR TITLE
[tests] Fix exit code check

### PR DESCRIPTION
**Pull request by ayushkovskiy**
Fix of exit-code check introduced in https://github.com/neuromation/cookiecutter-neuro-project/pull/175. Will close https://github.com/neuromation/cookiecutter-neuro-project/issues/162

### CODE REVIEW

**Comment by ayushkovskiy** [comment]
@anayden valid test on `make setup` fails with exit code 1: https://circleci.com/gh/neuromation/cookiecutter-neuro-project/855. Most likely, the problem is with pexpect's API we use.

Please take finish PR as it's intended to fix bug introduced in https://github.com/neuromation/cookiecutter-neuro-project/pull/175.

File: tests/e2e/helpers/runners.py [hunk 1]
```diff
@@ -197,7 +197,7 @@ def _run_once(
             finally:
                 chunk = _get_chunk(child)
                 output += chunk
-        if allow_nonzero_exitcode:
+        if not allow_nonzero_exitcode:
```
Comment (ayushkovskiy) [inline tests/e2e/helpers/runners.py]: I'd also rename this parameter to more clear `assert_exit_code`

File: tests/e2e/helpers/runners.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def run(
                 verbose=verbose,
                 detect_new_jobs=detect_new_jobs,
                 timeout_s=timeout_s,
-                allow_nonzero_exitcode=False,
+                allow_nonzero_exitcode=allow_nonzero_exitcode,
```
Comment (ayushkovskiy) [inline tests/e2e/helpers/runners.py]: another bug ^

**Comment by ayushkovskiy** [comment]
needs rebase on the newest master so that stage `cleanup e2e` does not hang

File: tests/e2e/helpers/runners.py [hunk 3]
```diff
@@ -197,8 +197,12 @@ def _run_once(
             finally:
                 chunk = _get_chunk(child)
                 output += chunk
-        if allow_nonzero_exitcode:
-            child.close()
+        if assert_exit_code:
+            if child.isalive():
+                # Short sleep to allow make command to complete
+                log_msg(f"Sleeping for {cmd}", logger=LOGGER.info)
+                time.sleep(1)
```
Comment (ayushkovskiy) [inline tests/e2e/helpers/runners.py]: what's the reason for this code? please provide either explanation/proof that we need it or let's drop it.

Comment (ayushkovskiy) [reply tests/e2e/helpers/runners.py]: and also explanation of this exact sleep delay.

Comment (anayden) [reply tests/e2e/helpers/runners.py]: I thought the comment above these lines was clear, my bad.

My experiments indicate that after we find all required matching patterns it takes a tiny bit of time for `make` process to complete. If we call `close()` immediately, the process gets killed and hence returns a non-zero `signalstatus`.

As far as _exact sleep delay is concerned_, I'm afraid there's no such thing in asynchronous world. There's only good enough and not good enough delays. This one proved to be just fine in all my tests.

Comment (ayushkovskiy) [reply tests/e2e/helpers/runners.py]: what I know is that any sleep is a potential source of flakiness, this is my concern. As an alternative, can we use `child.close(force=False)`? Will it wait until the child is closed?

Comment (anayden) [reply tests/e2e/helpers/runners.py]: > any sleep is a potential source of flakiness

I'm very well aware of that and agree with you. `close(force=False)` still sends `SIGTERM` to process which makes `signalstatus` non-zero, I tested that as well.

Comment (ayushkovskiy) [reply tests/e2e/helpers/runners.py]: should we use [`wait`](https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.wait) instead?

Comment (anayden) [reply tests/e2e/helpers/runners.py]: Already updated. Now dealing with NFS flakiness in an unrelated test

Comment (anayden) [reply tests/e2e/helpers/runners.py]: All tests pass, yay!

File: tests/e2e/helpers/runners.py [hunk 4]
```diff
@@ -197,12 +197,18 @@ def _run_once(
             finally:
                 chunk = _get_chunk(child)
                 output += chunk
-        if allow_nonzero_exitcode:
-            child.close()
+        if assert_exit_code:
+            if child.isalive():
+                # flush process buffer
+                child.read()
```
Comment (ayushkovskiy) [inline tests/e2e/helpers/runners.py]: let's save this last string: `output += child.read()`

File: tests/e2e/helpers/runners.py [hunk 5]
```diff
@@ -197,12 +197,18 @@ def _run_once(
...
+                # wait for child to exit
+                log_msg(f"Waiting for {cmd}", logger=LOGGER.info)
+                child.wait()
+            child.close(force=True)
```
Comment (ayushkovskiy) [inline tests/e2e/helpers/runners.py]: thank you, it seems to work! the last thing: let's add a few doctests here covering following situation:
1. `cmd="echo 1; sleep 2; echo 2", expected_patterns=["1"]`, so we check that the job's exit code is `0` and its output contains `2`
2. same `cmd`, `expected_patterns=[]`
3. `cmd="echo 1"` -- we check the branch where `child.isalive()` is False

Comment (ayushkovskiy) [reply tests/e2e/helpers/runners.py]: @anayden can we have these changes tested?

Comment (ayushkovskiy) [reply tests/e2e/helpers/runners.py]: We need this functionality. So let's merge now and add tests later: https://github.com/neuromation/cookiecutter-neuro-project/issues/230

File: tests/e2e/helpers/runners.py (hunk 4 above)
Comment (ayushkovskiy) [reply tests/e2e/helpers/runners.py]: fixed.

**Comment by ayushkovskiy** [comment]
tests fixed and actualized with changes in master.

**Pull request merged**